# Appendix {.unnumbered}

## Word count

The total word count on the main text is 4998:

![](figures/wordcount.png){width=40%}

## Suplementary figures

![Barplot showing the total number of sequenced reads (in millions) per sample, with bars coloured by experimental condition.](figures/lib_sizes.pdf){#suppfig-libsize}

![Boxplots of gene expression values by sample, coloured by experimental condition and faceted by transformation method. Note that CPM* denotes normalisation only, without variance stabilisation.](figures/transformed_filter_boxplot.pdf){#suppfig-boxplots}

![Mean versus standard deviation plot of gene expression values, faceted by transformation method. The red line indicates the mean-dependent trend in standard deviation. CPM* lacks variance stabilisation, as reflected in its stronger meanâ€“SD dependence.](figures/meansd_plot.pdf){#suppfig-meansd}

![Scree plot of the rlog-transformed data showing the number of retained principal components based on parallel analysis.](figures/screeplot_rlog.pdf){#suppfig-scree}

![Samples projected onto pairwise combinations of retained principal components from the rlog-transformed data, coloured by experimental condition.](figures/pca_rlog_plot.pdf){#suppfig-pca-rlog}

![Dispersion estimates plotted against mean expression levels. Gray points represent gene-wise dispersion estimates, blue points show shrinkage-based estimates, and the red line indicates the mean-dependent trend. Panel (A) corresponds to unfiltered data, while panel (B) shows filtered data.](figures/dispersion_plot.pdf){#suppfig-dispersion}

![Histograms of unadjusted p-values for each contrast.](figures/pvalue_hist.pdf){#suppfig-pvalue}

![MA plots displaying log2 fold change (LFC) estimates with shrinkage across mean expression levels for each contrast. Red points indicate significantly differentially expressed genes, while gray points represent non-significant genes.](figures/ma_plot.pdf){#suppfig-ma}

![Gap statistic plot showing the optimal number of clusters for the top 500 most variable genes based on rlog-transformed data. The optimal cluster number is indicated by the vertical dashed line.](figures/gap_plot.pdf){#suppfig-gap}