# Conclusions {#sec-conclusions}

This report presents the analysis of RNA-Seq data from an experiment investigating the response of *Cryptococcus neoformans* to cell wall stress across various environmental and genetic conditions, focusing on the FLC1 gene, a promising antifungal drug target. The study aimed to provide statistically robust insights into transcriptomic responses to combinations of growth media (CFW, EGTA, or both), temperature (30°C and 37°C), and the presence or absence of FLC1. The objectives were to assess replicate consistency, explore broad expression patterns among samples, identify differentially expressed genes across conditions, and reveal structured expression profiles within those genes.

Replicates clustered closely in PCA, indicating good biological consistency. Samples grown in YPD and YPD + CFW, as well as YPD + EGTA and YPD + CFW + EGTA, clustered together, but similarity varied by strain. In the FLC1Δ strain, adding CFW caused greater expression changes than in WT, with distinct profiles especially in YPD + CFW. Genes driving these patterns included CNAG_04283 (FLC1), CNAG_01653, CNAG_04891, and CNAG_00588. At 37°C, the FLC1Δ strain showed more expression variation across media, while at 30°C, greater overall divergence was observed. However, PCA loadings alone do not confirm differential expression for individual genes.

About 52.4% of genes (3,467) were differentially expressed in strain-media and strain-temperature contrasts relative to WT at 37°C in YPD. The greatest number of DE genes occurred in the FLC1Δ strain at 37°C with YPD + CFW + EGTA, mostly downregulated, with the highest downregulation (62%) seen in FLC1Δ at 37°C in YPD + CFW. Gene CNAG\_06576 was consistently significant, generally upregulated except for downregulation in that latter condition. Clustering the top 500 most variable DE genes identified four groups: cluster 1 genes were expressed above average in WT and below average in FLC1Δ, while clusters 2–4 showed the opposite pattern, indicating a strain-specific expression shift. Cluster 1 genes had higher expression under basal conditions (YPD, 30°C) and lower under stress (CFW, EGTA, 37°C), with clusters 2–4 displaying inverse trends more marked in FLC1Δ. Notably, cluster 4 genes decreased expression with EGTA and were largely unaffected by temperature. In addition, a ranked list of the top 100 DE genes with cluster annotations is provided for future study.

One limitation of this analysis is the reliance on a single algorithm for detecting differentially expressed genes, specifically the method implemented in the `DESeq2` package. The robustness of the findings presented in this report could be further evaluated by applying alternative differential expression analysis methods.



